### PR TITLE
[onert] Add quant8 support for Pad and PadV2 cker

### PR DESCRIPTION
**Pull request by hasw7569**
Add quant8 support for Pad and PadV2 cker

ONE-DCO-1.0-Signed-off-by: Leehansol <hansolu.lee@samsung.com>

### CODE REVIEW

File: runtime/onert/backend/cpu/ops/PadLayer.cc [hunk 1]
```diff
@@ -55,11 +54,11 @@ void PadLayer::run()
 {
   if (_input->data_type() == OperandType::FLOAT32)
   {
-    padFloat32();
+    PadImpl<float, float>(_constantValueData.f);
```
Comment (dr-venkman) [inline runtime/onert/backend/cpu/ops/PadLayer.cc:57]: I wonder whether we strictly need two typenames here? The padded constant values are likely to be of the same datatype as the input and output tensors.

Comment (hseok-oh) [reply runtime/onert/backend/cpu/ops/PadLayer.cc:57]: As @dr-venkman commented, input, output and constant values are same datatype. Tensorflow lite kernel check type for these

```cpp
  PadContext op_context(context, node);
  TF_LITE_ENSURE_TYPES_EQ(context, op_context.input->type,
                          op_context.output->type);
  if (op_context.constant_values != nullptr) {
    TF_LITE_ENSURE_TYPES_EQ(context, op_context.input->type,
                            op_context.constant_values->type);
  }
```

Comment (hasw7569) [reply runtime/onert/backend/cpu/ops/PadLayer.cc]: -> I wonder whether we strictly need two typenames here. The padded constant values are likely to be of the same datatype as the input and output tensors.

I agree.
I fixed it.

File: runtime/onert/backend/cpu/ops/PadLayer.cc [hunk 2]
```diff
@@ -27,39 +27,44 @@ namespace cpu
...
   _padRank = padRank;
-  _constantValueData.u8 = constantValueData;
+  if (_input->data_type() == OperandType::FLOAT32)
+  {
+    _value.f = constantValueData ? *reinterpret_cast<const float *>(constantValueData) : 0.0f;
+  }
+  else if (_input->data_type() == OperandType::QUANT_UINT8_ASYMM)
+  {
+    _value.u8 = constantValueData ? *reinterpret_cast<const uint8_t *>(constantValueData) : 0;
+  }
```
Comment (hseok-oh) [inline runtime/onert/backend/cpu/ops/PadLayer.cc:56]: Is there any reason to define new `ConstantValue` union and change implementation?

File: runtime/onert/backend/cpu/ops/PadLayer.h [hunk 3]
```diff
@@ -39,12 +39,10 @@ class PadLayer : public ::onert::exec::IFunction
   PadLayer();
 
 public:
-  void padFloat32();
-
-  void padQuant8();
+  template <typename T> void PadImpl(const T *constant_value_data);
```
Comment (hseok-oh) [inline runtime/onert/backend/cpu/ops/PadLayer.h:42]: Please use small case for method's first spell.

File: runtime/onert/backend/cpu/ops/PadLayer.cc [hunk 4]
```diff
@@ -27,39 +27,44 @@ namespace cpu
...
   if (_input->data_type() == OperandType::FLOAT32)
   {
-    padFloat32();
+    PadImpl<float>(static_cast<const float *>(&_value.f));
   }
   else if (_input->data_type() == OperandType::QUANT_UINT8_ASYMM)
   {
-    padQuant8();
+    PadImpl<uint8_t>(static_cast<const uint8_t *>(&_value.u8));
   }
```
Comment (hseok-oh) [inline runtime/onert/backend/cpu/ops/PadLayer.cc:68]: By using above my suggestion, we don't need casting

```cpp
  if (_input->data_type() == OperandType::FLOAT32)
  {
    padImpl<float>(_constantValueData.f);
  }
  else if (_input->data_type() == OperandType::QUANT_UINT8_ASYMM)
  {
    padImpl<uint8_t>(_constantValueData.u8);
  }
```

File: runtime/onert/backend/cpu/ops/PadLayer.cc (hunk 2 above)
Comment (hasw7569) [reply runtime/onert/backend/cpu/ops/PadLayer.cc:56]: In `KernelGenerator.cc` we can read input value to use for PadV2 by `ctx.at(index).data()->base()`.
but it always return const pointer.

In contrast, the 'Dataptr' union used previously is not a const type pointer, so I had a problem between type conversions.
So I define new `ConstantValue` to solve this problem.

Comment (hasw7569) [reply runtime/onert/backend/cpu/ops/PadLayer.cc:56]: > If you want to save const pointer, you can define ConstDataPtr uinon in OperationUtils.h and change _constantValueData type to ConstDataPtr. Then we need small change for this implementation.

good suggestion.
let me fix it.

File: runtime/onert/backend/cpu/ops/PadLayer.h (hunk 3 above)
Comment (hasw7569) [reply runtime/onert/backend/cpu/ops/PadLayer.h:42]: Oops.
I will fix it.

File: runtime/onert/backend/cpu/ops/PadLayer.cc [hunk 5]
```diff
@@ -33,33 +33,32 @@ PadLayer::PadLayer()
...
 {
   if (_input->data_type() == OperandType::FLOAT32)
   {
-    padFloat32();
+    padImpl<float>(_constantValueData.f);
   }
   else if (_input->data_type() == OperandType::QUANT_UINT8_ASYMM)
   {
-    padQuant8();
+    padImpl<uint8_t>(_constantValueData.u8);
```
Comment (hseok-oh) [inline runtime/onert/backend/cpu/ops/PadLayer.cc:61]: I checked `Pad` tflite kernel and found out default padding value for `QASYMM8` type is input's `offset`, not 0. So we should pass input's offset value

```cpp
if (_constantValueDate.u8 == nullptr)
{
  uint8_t pad_value = static_cast<uint8_t>(_input->data_offset());
  padImpl<uint8_t>(&pad_value);
}
else
{
  padImpl<uint8_t>(_constantValueData.u8);
}
```

File: tests/nnapi/specs/V1_1/pad_quant8_nnfw.mod.py [hunk 6]
```diff
@@ -11,10 +11,10 @@
...
 
 output0 = {i3: # output 0
-           [2, 2, 2, 2,
-            2, 1, 2, 2,
-            2, 3, 4, 2,
-            2, 2, 2, 2]}
+           [0, 0, 0, 0,
+            0, 1, 2, 0,
+            0, 3, 4, 0,
+            0, 0, 0, 0]}
```
Comment (hseok-oh) [inline tests/nnapi/specs/V1_1/pad_quant8_nnfw.mod.py:17]: Please revert this change. (refer https://github.com/Samsung/ONE/pull/3220/files#r458522865)

File: runtime/onert/backend/cpu/ops/PadLayer.cc (hunk 5 above)
Comment (hasw7569) [reply runtime/onert/backend/cpu/ops/PadLayer.cc]: oh..
I missed it.
I will fix it.

File: tests/nnapi/nnapi_gtest.skip.armv7l-linux.acl_neon [hunk 7]
```diff
@@ -111,6 +111,13 @@ GeneratedTests.not_equal_dynamic_float_nnfw
 GeneratedTests.one_hot_ex_dynamic_nnfw
 GeneratedTests.pack_ex_dynamic_nnfw
 GeneratedTests.pad_dynamic_nnfw
+GeneratedTests.pad_quant8_nnfw
```
Comment (hseok-oh) [inline tests/nnapi/nnapi_gtest.skip.armv7l-linux.acl_neon:114]: We can remove `GeneratedTests.pad_quant8_nnfw` from skiplit for armv7-linux acl_cl, acl_neon, aarch64-linux acl_cl, acl_neon,

**Comment by hasw7569** [comment]
@dr-venkman  PTAL

**Comment by hseok-oh** [comment]
@dr-venkman PTAL

File: runtime/onert/backend/cpu/ops/PadLayer.cc [hunk 8]
```diff
@@ -33,33 +33,40 @@ PadLayer::PadLayer()
...
   {
-    padFloat32();
+    padImpl<float>(_constantValueData.f);
   }
   else if (_input->data_type() == OperandType::QUANT_UINT8_ASYMM)
   {
-    padQuant8();
+    if (_constantValueData.u8 == nullptr)
+    {
+      uint8_t pad_value = static_cast<uint8_t>(_input->data_offset());
```
Comment (dr-venkman) [inline runtime/onert/backend/cpu/ops/PadLayer.cc:63]: Shouldn't it refer to `_output->data_offset()1 instead, looking from TFLite implementation? Please check and ignore the comment if unecessary.

Comment (hasw7569) [reply runtime/onert/backend/cpu/ops/PadLayer.cc:63]: I will fix it.
